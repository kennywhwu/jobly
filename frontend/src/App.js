import React, { Component } from 'react';
import jwt from 'jsonwebtoken';
import './App.css';
import NavBar from './NavBar';
import Routes from './Routes';
import Alert from './Alert';
import JoblyApi from './JoblyApi';

class App extends Component {
  constructor(props) {
    super(props);
    this.state = {
      currentUser: null,
      isLoading: true,
      alert: { type: null, message: '' }
    };
    this.fetchUser = this.fetchUser.bind(this);
    this.clearUser = this.clearUser.bind(this);
    this.triggerAlert = this.triggerAlert.bind(this);
  }

  async componentDidMount() {
    this.fetchUser();
  }

  async fetchUser() {
    try {
      let token = JSON.parse(localStorage.getItem('token'));
      if (token !== null) {
        let payload = jwt.decode(token);
        let currentUser = await JoblyApi.getUser(payload.username);
        let appliedJobsResult = await JoblyApi.getJobsUserAppliedTo(
          payload.username
        );
        let jobsAppliedTo = new Set();
        appliedJobsResult.forEach(job => jobsAppliedTo.add(job.job_id));
        currentUser.jobsAppliedTo = jobsAppliedTo;
        this.setState({ currentUser });
      }
    } catch (err) {
      this.triggerAlert('danger', err);
    }
    this.setState({ isLoading: false });
  }

  clearUser() {
    this.setState({ currentUser: null }, () => {
      let token = localStorage.getItem('token');
      localStorage.removeItem('token');
      if (token) {
        this.props.triggerAlert('success', 'Logged out!');
      }
    });
  }

  triggerAlert(type, message) {
    this.setState({ alert: { type, message } });
    setTimeout(
      () => this.setState({ alert: { type: null, message: '' } }),
      3000
    );
  }

  render() {
    const alert = this.state.alert;
    return (
      <div className="App">
        {!this.state.isLoading ? (
          <div>
            <NavBar currentUser={this.state.currentUser} />
            {alert.type !== null ? (
              <Alert type={alert.type} message={alert.message} />
            ) : (
              undefined
            )}
            <Routes
              currentUser={this.state.currentUser}
              fetchUser={this.fetchUser}
              clearUser={this.clearUser}
              triggerAlert={this.triggerAlert}
            />
          </div>
        ) : (
          <h1>Loading...</h1>
        )}
      </div>
    );
  }
}

export default App;
