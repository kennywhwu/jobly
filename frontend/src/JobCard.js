// JobCard component for individual jobs

import React, { Component } from "react";
import { Button, Card, CardBody, CardTitle, CardText } from "reactstrap";
import JoblyApi from "./JoblyApi";

class JobCard extends Component {
  constructor(props) {
    super(props);
    this.state = {
      applied: this.props.currentUser.jobsAppliedTo.has(this.props.job.id)
        ? true
        : false,
    };
    this.apply = this.apply.bind(this);
  }

  async apply() {
    let message = await JoblyApi.apply(this.props.job.id, this.props.job.state);
    if (message) {
      this.setState({ applied: message, disabled: true });
      this.props.triggerAlert("success", "Applied to job!");
      this.props.fetchUser();
    }
  }

  render() {
    return (
      <Card className="JobCard">
        <CardBody className="CardBody">
          <CardTitle className="company-name CardTitle">
            {this.props.job.title}
          </CardTitle>
          <CardText className="company-description CardText">
            Salary: {this.props.job.salary}
          </CardText>
          <CardText className="company-description CardText">
            Equity: {this.props.job.equity}
          </CardText>
          <Button
            color="primary"
            onClick={this.apply}
            disabled={this.state.applied ? true : false}
          >
            {this.state.applied ? "Applied" : "Apply"}
          </Button>
        </CardBody>
      </Card>
    );
  }
}

export default JobCard;
