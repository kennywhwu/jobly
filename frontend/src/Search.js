// Search form component

import React, { Component } from 'react';

// Filters list based on entered search term
class Search extends Component {
  constructor(props) {
    super(props);
    this.state = {
      name: ''
    };
    this.handleChange = this.handleChange.bind(this);
    this.handleSubmit = this.handleSubmit.bind(this);
  }

  // Handle change to user input boxes
  handleChange(evt) {
    this.setState({
      [evt.target.name]: evt.target.value
    });
  }

  // Handle submission of form
  handleSubmit(evt) {
    evt.preventDefault();
    this.props.handleSearch(this.state.name);
  }

  render() {
    return (
      <form onSubmit={this.handleSubmit}>
        <input
          type="text"
          name="name"
          id="name"
          onChange={this.handleChange}
          placeholder="Enter search term..."
        />
        <button>Submit</button>
      </form>
    );
  }
}

export default Search;
