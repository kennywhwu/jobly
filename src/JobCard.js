import React, { Component } from 'react';

class JobCard extends Component {
  render() {
    return (
      <div className="JobCard">
        <div>{this.props.job.title}</div>
        <div>Salary: {this.props.job.salary}</div>
        <div>Equity: {this.props.job.equity}</div>
        <button>Apply</button>
        <hr />
      </div>
    );
  }
}

export default JobCard;
