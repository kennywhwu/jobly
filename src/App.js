import React, { Component } from 'react';
import './App.css';
import NavBar from './NavBar';
import Routes from './Routes';

class App extends Component {
  render() {
    return (
      <div className="App">
        <NavBar />
        <Routes />
      </div>
    );
  }
}

export default App;
