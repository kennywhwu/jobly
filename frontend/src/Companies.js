// Companies component for list of companies

import React, { Component } from "react";
import { Button } from "reactstrap";
import JoblyApi from "./JoblyApi";
import CompanyCard from "./CompanyCard";
import Search from "./Search";

class Companies extends Component {
  constructor(props) {
    super(props);
    this.state = { companies: [], page: 0 };
    this.filterCompanies = this.filterCompanies.bind(this);
    this.changePage = this.changePage.bind(this);
  }

  static defaultProps = {
    itemsPerPage: 20,
  };

  // Get list of companies when component first mounts
  async componentDidMount() {
    let companyResults = await JoblyApi.getCompanies();
    this.setState({ companies: companyResults });
  }

  // Filter rendered companies list based on search term
  async filterCompanies(name) {
    let results = await JoblyApi.getCompaniesSearch(name);
    this.setState({ companies: results });
  }

  changePage(evt) {
    this.setState({ page: +evt.target.id });
  }

  render() {
    let companies = this.state.companies.slice(
      this.state.page * this.props.itemsPerPage,
      (1 + this.state.page) * this.props.itemsPerPage
    );
    let pages = [
      ...Array(
        Math.ceil(this.state.companies.length / this.props.itemsPerPage)
      ).keys(),
    ];

    return (
      <div className="Companies">
        <div className="header">
          <h3>Companies</h3>
          <Search handleSearch={this.filterCompanies} />
          {pages.map(page => (
            <Button
              onClick={this.changePage}
              id={page}
              className="button-page"
              key={page}
              disabled={this.state.page === page && true}
              color="info"
            >
              {page + 1}
            </Button>
          ))}
        </div>
        {companies.map(company => (
          <CompanyCard
            key={company.handle}
            company={company}
            triggerAlert={this.props.triggerAlert}
          />
        ))}
        {pages.map(page => (
          <Button
            onClick={this.changePage}
            id={page}
            className="button-page"
            key={page}
            disabled={this.state.page === page && true}
            color="info"
          >
            {page + 1}
          </Button>
        ))}
      </div>
    );
  }
}

export default Companies;
