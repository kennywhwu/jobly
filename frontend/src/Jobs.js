// Jobs component for list of jobs

import React, { Component } from 'react';
import JoblyApi from './JoblyApi';
import JobCard from './JobCard';
import Search from './Search';

class Jobs extends Component {
  constructor(props) {
    super(props);
    this.state = { jobs: [] };
    this.filterJobs = this.filterJobs.bind(this);
  }

  async componentDidMount() {
    // [{handle:, name:...},{},...]
    let results = await JoblyApi.getJobs();
    this.setState({ jobs: results });
  }

  async filterJobs(title) {
    let results = await JoblyApi.getJobsSearch(title);
    this.setState({ jobs: results });
  }

  render() {
    return (
      <div className="Jobs">
        Jobs
        <Search handleSearch={this.filterJobs} />
        {this.state.jobs.map(job => (
          <JobCard key={job.id} job={job} />
        ))}
      </div>
    );
  }
}

export default Jobs;
