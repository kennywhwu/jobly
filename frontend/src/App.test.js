import React from 'react';
import { shallow, mount } from 'enzyme';
import toJson from 'enzyme-to-json';
import { MemoryRouter } from 'react-router-dom';
import App from './App';
import Companies from './Companies';
import JobCard from './JobCard';
import Jobs from './Jobs';
import Search from './Search';

// Smoke test
it('renders shallowly without crashing', function() {
  shallow(<App />);
});

it('renders with routes without crashing', function() {
  mount(
    <MemoryRouter>
      <App />
    </MemoryRouter>
  );
});

it('renders Companies and Search components when going to /companies', function() {
  let wrapper = mount(
    <MemoryRouter initialEntries={['/companies']}>
      <App />
    </MemoryRouter>
  );

  let elem = wrapper.children();
  console.log('elem', elem.debug());
  expect(elem.matchesElement(<Companies />)).toBe(true);
  expect(elem.matchesElement(<Search />));
});
it('renders JobCard component when going to /companies/:handle', function() {
  let wrapper = mount(
    <MemoryRouter initialEntries={['/companies/baker-santos']}>
      <App />
    </MemoryRouter>
  );

  let elem = wrapper.find('div').first();
  expect(elem.matchesElement(<JobCard />));
});

it('renders Jobs and Search components when going to /jobs', function() {
  let wrapper = mount(
    <MemoryRouter initialEntries={['/jobs']}>
      <App />
    </MemoryRouter>
  );

  let elem = wrapper.find('div').first();
  expect(elem.matchesElement(<Jobs />));
  expect(elem.matchesElement(<Search />));
});
