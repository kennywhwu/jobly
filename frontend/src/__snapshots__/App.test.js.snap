// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders without crashing 1`] = `
<MemoryRouter
  initialEntries={
    Array [
      "/companies",
    ]
  }
>
  <Router
    history={
      Object {
        "action": "POP",
        "block": [Function],
        "canGo": [Function],
        "createHref": [Function],
        "entries": Array [
          Object {
            "hash": "",
            "key": "6epzoq",
            "pathname": "/companies",
            "search": "",
            "state": undefined,
          },
        ],
        "go": [Function],
        "goBack": [Function],
        "goForward": [Function],
        "index": 0,
        "length": 1,
        "listen": [Function],
        "location": Object {
          "hash": "",
          "key": "6epzoq",
          "pathname": "/companies",
          "search": "",
          "state": undefined,
        },
        "push": [Function],
        "replace": [Function],
      }
    }
  >
    <App>
      <div
        className="App"
      >
        <NavBar>
          <nav>
            <NavLink
              activeClassName="active"
              aria-current="page"
              to="/companies"
            >
              <Route
                path="\\\\/companies"
              >
                <Link
                  aria-current="page"
                  className="active"
                  replace={false}
                  style={Object {}}
                  to="/companies"
                >
                  <a
                    aria-current="page"
                    className="active"
                    href="/companies"
                    onClick={[Function]}
                    style={Object {}}
                  >
                    Companies
                  </a>
                </Link>
              </Route>
            </NavLink>
            <NavLink
              activeClassName="active"
              aria-current="page"
              to="/jobs"
            >
              <Route
                path="\\\\/jobs"
              >
                <Link
                  aria-current={null}
                  replace={false}
                  to="/jobs"
                >
                  <a
                    aria-current={null}
                    href="/jobs"
                    onClick={[Function]}
                  >
                    Jobs
                  </a>
                </Link>
              </Route>
            </NavLink>
            <NavLink
              activeClassName="active"
              aria-current="page"
              to="/profile"
            >
              <Route
                path="\\\\/profile"
              >
                <Link
                  aria-current={null}
                  replace={false}
                  to="/profile"
                >
                  <a
                    aria-current={null}
                    href="/profile"
                    onClick={[Function]}
                  >
                    Profile
                  </a>
                </Link>
              </Route>
            </NavLink>
            <NavLink
              activeClassName="active"
              aria-current="page"
              to="/"
            >
              <Route
                path="\\\\/"
              >
                <Link
                  aria-current="page"
                  className="active"
                  replace={false}
                  style={Object {}}
                  to="/"
                >
                  <a
                    aria-current="page"
                    className="active"
                    href="/"
                    onClick={[Function]}
                    style={Object {}}
                  >
                    Logout
                  </a>
                </Link>
              </Route>
            </NavLink>
          </nav>
        </NavBar>
        <Routes>
          <Switch>
            <Route
              computedMatch={
                Object {
                  "isExact": true,
                  "params": Object {},
                  "path": "/companies",
                  "url": "/companies",
                }
              }
              exact={true}
              location={
                Object {
                  "hash": "",
                  "key": "6epzoq",
                  "pathname": "/companies",
                  "search": "",
                  "state": undefined,
                }
              }
              path="/companies"
              render={[Function]}
            >
              <Companies>
                <div
                  className="Companies"
                >
                  Companies
                  <Search
                    handleSearch={[Function]}
                  >
                    <form
                      onSubmit={[Function]}
                    >
                      <input
                        id="name"
                        name="name"
                        onChange={[Function]}
                        placeholder="Enter search term..."
                        type="text"
                      />
                      <button>
                        Submit
                      </button>
                    </form>
                  </Search>
                </div>
              </Companies>
            </Route>
          </Switch>
        </Routes>
      </div>
    </App>
  </Router>
</MemoryRouter>
`;
